discoveryservice/module.go

```diff
@@ -0,0 +1,232 @@
+var _ core.Configurable = &Module{}
+var _ Server = &Module{}
+
+// var _ Client = &Module{}
+var retractionPresentationType = ssi.MustParseURI("RetractedVerifiablePresentation")
+
+func New(storageInstance storage.Engine) *Module {
+	return &Module{
+		storageInstance: storageInstance,
+	}
+}
+
+type Module struct {
+	config            Config
+	storageInstance   storage.Engine
+	store             *sqlStore
+	serverDefinitions map[string]Definition
+	services          map[string]Definition
+}
+
+func (m *Module) Configure(_ core.ServerConfig) error {
+	if m.config.Definitions.Directory == "" {
+		return nil
+	}
+	var err error
+	m.services, err = loadDefinitions(m.config.Definitions.Directory)
+	if err != nil {
+		return err
+	}
+	if len(m.config.Server.DefinitionIDs) > 0 {
+		// Get the definitions that are enabled for this server
+		serverDefinitions := make(map[string]Definition)
+		for _, definitionID := range m.config.Server.DefinitionIDs {
+			if definition, exists := m.services[definitionID]; !exists {
+				return fmt.Errorf("definition '%s' not found", definitionID)
+			} else {
+				serverDefinitions[definitionID] = definition
+			}
+		}
+		m.serverDefinitions = serverDefinitions
+	}
+	return nil
+}
+
+func (m *Module) Start() error {
+	var err error
+	m.store, err = newSQLStore(m.storageInstance.GetSQLDatabase(), m.services)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (m *Module) Shutdown() error {
+	return nil
+}
+
+func (m *Module) Name() string {
+	return ModuleName
+}
+
+func (m *Module) Config() interface{} {
+	return &m.config
+}
+
+func (m *Module) Add(serviceID string, presentation vc.VerifiablePresentation) error {
+	definition, exists := m.services[serviceID]
+	if !exists {
+		return ErrServiceNotFound
+	}
+	if _, isMaintainer := m.serverDefinitions[serviceID]; !isMaintainer {
+		return ErrServerModeDisabled
+	}
+	if presentation.Format() != vc.JWTPresentationProofFormat {
+		return errors.New("only JWT presentations are supported")
+	}
+	// TODO: validate signature
+	if presentation.ID == nil {
+		return errors.New("presentation does not have an ID")
+	}
+	expiration := presentation.JWT().Expiration()
+	// VPs should not be valid for too short; unnecessary overhead at the server and clients.
+	// Also protects against expiration not being set at all. The factor is somewhat arbitrary.
+	minValidity := float64(definition.PresentationMaxValidity / 4)
+	if expiration.Sub(time.Now()).Seconds() < minValidity {
+		return fmt.Errorf("presentation is not valid for long enough (min %s)", time.Duration(minValidity)*time.Second)
+	}
+	// VPs should not be valid for too long, as that would prevent the server from pruning them.
+	if int(expiration.Sub(time.Now()).Seconds()) > definition.PresentationMaxValidity {
+		return fmt.Errorf("presentation is valid for too long (max %s)", time.Duration(definition.PresentationMaxValidity)*time.Second)
+	}
+
+	if presentation.IsType(retractionPresentationType) {
```

Comment:
retract is till a weird word, what was wrong with revoke?

Reply:
I had "delete", but you don't really "delete" a presentation, since new a presentation just replaces the previous one.
"Revoke" is tightly associated with revoking credentials, a feature which we also have to implement (checking whether the presentation contains VC(s) that are revoked), so that becomes very confusing.

Prefer "delete" or "remove"?

Reply:
or "unsubscribe", "deregister". All still not very catchy.... dunno 🤷